tpm2/structures.go

```diff
@@ -137,6 +138,49 @@ func (p Public) Name() (*Name, error) {
 	}, nil
 }
 
+// MatchesTemplate checks if the Public area has the same algorithms and
+// parameters as the provided template. Note that this does not necessarily
+// mean that the key was created from this template, as the Unique field is
+// both provided in the template and overriden in the key creation process.
+func (p Public) MatchesTemplate(template Public) bool {
+	if p.Type != template.Type ||
+		p.NameAlg != template.NameAlg ||
+		p.Attributes != template.Attributes ||
+		!bytes.Equal(p.AuthPolicy, template.AuthPolicy) {
+		return false
+	}
+	switch p.Type {
+	case AlgRSA:
+		pParams := p.RSAParameters
+		tParams := template.RSAParameters
+		pExponent := pParams.Exponent
+		if pExponent == 0 {
+			pExponent = defaultRSAExponent
+		}
+		tExponent := tParams.Exponent
+		if tExponent == 0 {
+			tExponent = defaultRSAExponent
+		}
+		return reflect.DeepEqual(pParams.Symmetric, tParams.Symmetric) &&
+			reflect.DeepEqual(pParams.Sign, tParams.Sign) &&
```

Comment:
iirc these are fixed-size structs, can we compare them directly instead?

Reply:
The main reason this doesn't work is when we are dealing with pointer types. We have `var a *Foo` and `var b *Foo`, and we want them to compare equal if they are either both `nil`, or their field values compare equal. Unfortunately, there's not an easy way to do this in go.

Comparing like `a == b` will check if they point to the same object (not what we want), and `*a == *b` risks a segfault on `nil` pointers. Unfortunately, it seems like `reflect.DeepEqual` is the easiest wa

Reply:
SGTM